libtenzir/builtins/operators/sample.cpp

```diff
@@ -0,0 +1,214 @@
+        case mode::log2:
+          return std::log2(count);
+        case mode::log10:
+          return std::log10(count);
+        case mode::sqrt:
+          return std::sqrt(count);
+        case mode::none:
+          TENZIR_UNREACHABLE();
+      }
+      TENZIR_UNREACHABLE();
+    };
+    constexpr auto make_stride_index
+      = [](const int64_t offset, const int64_t rows, const int64_t stride) {
+          TENZIR_ASSERT(stride > 0);
+          auto b = int64_type::make_arrow_builder(arrow::default_memory_pool());
+          check(b->Reserve((rows + 1) / stride));
+          for (auto i = offset % stride; i < rows; i += stride) {
+            check(b->Append(i));
+          }
+          return finish(*b);
+        };
+    for (const auto& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      // XXX: >=?
+      if (auto now = std::chrono::steady_clock::now(); now - last > delta_) {
+        stride = std::ceil(calc());
+        last = now;
+        offset = 0;
+        count = 0;
+      }
+      count += slice.rows();
+      auto batch = to_record_batch(slice);
+      const auto rows = batch->num_rows();
+      auto stride_index = make_stride_index(offset, rows, stride);
+      offset += rows;
+      auto take_result = arrow::compute::Take(batch, stride_index);
+      if (not take_result.ok()) {
+        // XXX: Should this be an error?
+        diagnostic::error("{}", take_result.status().ToString())
+          .note("failed to apply stride")
+          .throw_();
+      }
+      const auto datum = take_result.MoveValueUnsafe();
+      TENZIR_ASSERT(datum.kind() == arrow::Datum::Kind::RECORD_BATCH);
+      co_yield table_slice{
+        datum.record_batch(),
+        slice.schema(),
+      };
+    }
+  }
+
+  auto name() const -> std::string override {
+    return "tql2.sample";
+  }
+
+  auto location() const -> operator_location override {
+    // TODO: Figure this out
+    return operator_location::local;
+  }
+
+  auto
+  optimize(expression const&, event_order) const -> optimize_result override {
+    // TODO: Figure this out
+    return do_not_optimize(*this);
+  }
+
+  friend auto inspect(auto& f, sample_operator& x) -> bool {
+    return f.object(x)
+      .pretty_name("tenzir.plugins.sample.sample_operator")
+      // FIXME: Add mode_
+      .fields(f.field("delta", x.delta_));
+  }
+
+private:
+  mode mode_;
+  duration delta_;
+};
+
+class plugin final : public virtual operator_plugin<sample_operator>,
+                     public virtual operator_factory_plugin {
+public:
+  auto signature() const -> operator_signature override {
+    return {.transformation = true};
+  }
+
+  auto
+  make(invocation inv, session ctx) const -> failure_or<operator_ptr> override {
+    auto mode = located<std::string>{};
+    auto delta = located<duration>{};
+    // TODO: Decide name
+    TRY(argument_parser2::operator_("sample")
+          .add("mode", mode)
+          .add("delta", delta)
```

Comment:
You mentioned `window` as a new name in a call; I agree that works better.

We can also continue to be inspired and name it `period` and also consider adding `min_events` and `max_rate` arguments.